### PR TITLE
Adding custom client loggers and metrics

### PR DESCRIPTION
**Pull request by ravirajj**

### CODE REVIEW

File: runtime/call_metrics.go [hunk 1]
```diff
@@ -234,3 +243,54 @@ func NewOutboundTChannelMetrics(scope tally.Scope) *OutboundTChannelMetrics {
...
+	clientID string,
+	methods []string,
+	deps *DefaultDependencies,
+) (
+	logger *zap.Logger,
+	scope tally.Scope,
+	loggers map[string]*zap.Logger,
+	metrics map[string]*OutboundCustomMetrics,
+) {
+	logger = deps.Logger.With(zap.String("clientID", clientID))
```
Comment (cl1337) [inline runtime/call_metrics.go]: just curious why `clientID -> clientID` and `client -> clientID` :)

File: runtime/call_metrics.go [hunk 2]
```diff
@@ -234,3 +243,54 @@ func NewOutboundTChannelMetrics(scope tally.Scope) *OutboundTChannelMetrics {
...
+	for _, methodName := range methods {
+		loggers[methodName] = logger.With(
+			zap.String("methodName", methodName),
+		)
+		metrics[methodName] = NewOutboundCustomMetrics(scope.Tagged(map[string]string{
+			"method": methodName,
+		}))
+	}
+
+	return logger, scope, loggers, metrics
```
Comment (cl1337) [inline runtime/call_metrics.go]: why T_T

File: examples/example-gateway/clients/quux/quux.go [hunk 3]
```diff
@@ -21,18 +21,43 @@
 package quux
 
 import (
-	module "github.com/uber/zanzibar/examples/example-gateway/build/clients/quux/module"
+	"time"
+
+	"github.com/uber/zanzibar/examples/example-gateway/build/clients/quux/module"
 	zanzibar "github.com/uber/zanzibar/runtime"
+	"go.uber.org/zap"
```
Comment (cl1337) [inline examples/example-gateway/clients/quux/quux.go]: nit: space before local pkg & 3rd pkg

Comment (ravirajj) [reply examples/example-gateway/clients/quux/quux.go]: Done.

File: runtime/call_metrics.go (hunk 1 above)
Comment (ravirajj) [reply runtime/call_metrics.go]: Originally, I had `client` as a field and tag key, but was asked to follow this convention of adding `ID` or `Name` to the key for logger fields.

File: runtime/call_metrics.go (hunk 2 above)
Comment (ravirajj) [reply runtime/call_metrics.go]: Why what? Not sure what your question is.

File: runtime/call_metrics.go [hunk 4]
```diff
@@ -234,3 +243,54 @@ func NewOutboundTChannelMetrics(scope tally.Scope) *OutboundTChannelMetrics {
 	metrics.SystemErrors = scope.Counter(outboundCallsSystemErrors)
 	return &metrics
 }
+
+// NewOutboundCustomMetrics returns outbound metrics for custom clients
+func NewOutboundCustomMetrics(scope tally.Scope) *OutboundCustomMetrics {
```
Comment (Raynos) [inline runtime/call_metrics.go]: Custom metrics do not belong here. Metrics integration belongs at the http client or tchannel client level.

File: examples/example-gateway/clients/quux/quux.go [hunk 5]
```diff
@@ -21,18 +21,44 @@
...
+type quux struct {
+	loggers map[string]*zap.Logger
+	metrics map[string]*zanzibar.OutboundCustomMetrics
+}
 
-// NewClient creates a new Quux client
+// NewClient creates a new Quux client, with client method loggers and metrics.
 func NewClient(g *zanzibar.Gateway, deps *module.Dependencies) Client {
-	return &quux{}
+	logger, scope, loggers, metrics := zanzibar.SetupCustomClientLoggersMetrics(
```
Comment (Raynos) [inline examples/example-gateway/clients/quux/quux.go]: We shouldn't do custom metrics.

Custom loggers is fine so `zanzibar.SetupClientLoggers(name, methods, deps.Default)` is fine.

**Pull request closed without merging**